### PR TITLE
injector: exit when ListenAndServeTLS fails and prevent too many timers

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

File: pilot/pkg/kube/inject/webhook.go [hunk 1]
```diff
@@ -226,7 +233,7 @@ func (wh *Webhook) Run(stop <-chan struct{}) {
 			wh.mu.Unlock()
 		case event := <-wh.watcher.Event:
 			// use a timer to debounce configuration updates
-			if event.IsModify() || event.IsCreate() {
+			if (event.IsModify() || event.IsCreate()) && timerC == nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/kube/inject/webhook.go]: for reviewers: this is to prevent to start too many timers.

File: pilot/pkg/kube/inject/webhook.go [hunk 2]
```diff
@@ -188,12 +188,18 @@ func NewWebhook(p WebhookParameters) (*Webhook, error) {
...
+		case <-stop:
+			log.Infof(msg)
+		default:
+			log.Fatalf("%s due to error: %v", msg, err)
 		}
 	}()
-	defer wh.watcher.Close() // nolint: errcheck
-	defer wh.server.Close()  // nolint: errcheck
+	defer wh.watcher.Close()
+	defer wh.server.Close()
```
Comment (ayj) [inline pilot/pkg/kube/inject/webhook.go]: Does this no longer cause a linter error?

File: pilot/pkg/kube/inject/webhook.go [hunk 3]
```diff
@@ -188,12 +188,18 @@ func NewWebhook(p WebhookParameters) (*Webhook, error) {
...
-		if err := wh.server.ListenAndServeTLS("", ""); err != nil {
-			log.Errorf("ListenAndServeTLS for admission webhook returned error: %v", err)
+		err := wh.server.ListenAndServeTLS("", "")
+
+		msg := fmt.Sprintf("Stopped listening on %s", wh.server.Addr)
+		select {
+		case <-stop:
+			log.Infof(msg)
+		default:
+			log.Fatalf("%s due to error: %v", msg, err)
```
Comment (ayj) [inline pilot/pkg/kube/inject/webhook.go]: Can we copy the approach taken by the validation webhook? i.e.
https://github.com/istio/istio/blob/efd8e8478248dca2cf0c15c908583552adc235be/galley/pkg/crd/validation/webhook.go#L211-L215

Comment (hzxuzhonghu) [reply pilot/pkg/kube/inject/webhook.go]: Yes, we can. Will do that.

File: pilot/pkg/kube/inject/webhook.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/kube/inject/webhook.go]: No, it doesn't now.

**Comment by hzxuzhonghu** [comment]
@ayj addressed.

**Comment by rshriram** [comment]
you can't add lgtm label :).. the bot doesn't look at that unfortunately.

**Comment by hzxuzhonghu** [comment]
rebased to fix circleci

**Pull request merged**